### ISSUE TITLE
Scipy.optimize.minimize documentation: args specified as tuple

### ISSUE THREAD

**Issue by hermanc1** [scipy/scipy#6757]
In the documentation for scipy.optimize.minimize, the `args` parameter is specified as tuple. I think it should be a dictionary. At least, I can get a dictionary to work, but not a tuple.

Illustration from docs:

    import scipy.optimize.minimize
    print(scipy.optimize.minimize.__doc__)

    Minimization of scalar function of one or more variables.
    
    In general, the optimization problems are of the form::
    
        minimize f(x) subject to

        g_i(x) >= 0,  i = 1,...,m
        h_j(x)  = 0,  j = 1,...,p

    where x is a vector of one or more variables.
    ``g_i(x)`` are the inequality constraints.
    ``h_j(x)`` are the equality constrains.

    Optionally, the lower and upper bounds for each element in x can also be
    specified using the `bounds` argument.

    Parameters
    ----------
    fun : callable
        Objective function.
    x0 : ndarray
        Initial guess.
    args : tuple, optional
        Extra arguments passed to the objective function and its
        derivatives (Jacobian, Hessian).`

**Comment by andyfaff** [comment]
The extra arguments supplied to the objective function are passed in through a tuple, at least in all the code I've ever written that uses `minimize`. This seems to be more of a support question, so you're probably going to be better off asking this question on the mailing list. It's a good idea to have a small self contained example to illustrate the problem. If it is a bug, then please reopen this issue.

**Comment by hermanc1** [comment]
It is entirely possible that I'm just being a noob and not specifying the tuple correctly, but here is an example. Specifying as a tuple doesn't work:

    import numpy as np
    import scipy.stats

    def sse(params, args):
        data, psr = args[0], args[1]
        #data, psr = args['dn'], args['r']
        betavar, rho, zeta = params[0], params[1], params[2]
        return sum((data - (rho + zeta*np.exp(betavar*psr))**-1)**2)

    past_sample_range = [0, 1, 2]
    data_known = [0.98, 0.98, 0.92]
    additional = (data_known, range(len(past_sample_range)))
    #additional = {'dn': data_known, 'r':range(len(past_sample_range))}
    res = scipy.optimize.minimize(fun=sse, args=additional, x0=np.array([0.1, 0.0001, 1]), bounds = ((0, None), (0, None), (0, None)))

Which returns:
```
Traceback (most recent call last):

  File "<ipython-input-14-d744c9d6f05c>", line 1, in <module>
    runfile('/home/herman/.config/spyder-py3/temp.py', wdir='/home/herman/.config/spyder-py3')

  File "/home/herman/anaconda3/lib/python3.5/site-packages/spyder/utils/site/sitecustomize.py", line 866, in runfile
    execfile(filename, namespace)

  File "/home/herman/anaconda3/lib/python3.5/site-packages/spyder/utils/site/sitecustomize.py", line 102, in execfile
    exec(compile(f.read(), filename, 'exec'), namespace)

  File "/home/herman/.config/spyder-py3/temp.py", line 20, in <module>
    res = scipy.optimize.minimize(fun=sse, args=additional, x0=np.array([0.1, 0.0001, 1]), bounds = ((0, None), (0, None), (0, None)))

  File "/home/herman/anaconda3/lib/python3.5/site-packages/scipy/optimize/_minimize.py", line 450, in minimize
    callback=callback, **options)

  File "/home/herman/anaconda3/lib/python3.5/site-packages/scipy/optimize/lbfgsb.py", line 328, in _minimize_lbfgsb
    f, g = func_and_grad(x)

  File "/home/herman/anaconda3/lib/python3.5/site-packages/scipy/optimize/lbfgsb.py", line 273, in func_and_grad
    f = fun(x, *args)

  File "/home/herman/anaconda3/lib/python3.5/site-packages/scipy/optimize/optimize.py", line 292, in function_wrapper
    return function(*(wrapper_args + args))

TypeError: sse() takes 2 positional arguments but 3 were given
```

But specifying a dict works:

    import numpy as np
    import scipy.stats

    def sse(params, args):
        #data, psr = args[0], args[1]
        data, psr = args['dn'], args['r']
        betavar, rho, zeta = params[0], params[1], params[2]
        return sum((data - (rho + zeta*np.exp(betavar*psr))**-1)**2)

    past_sample_range = [0, 1, 2]
    data_known = [0.98, 0.98, 0.92]
    #additional = (data_known, range(len(past_sample_range)))
    additional = {'dn': data_known, 'r':range(len(past_sample_range))}
    res = scipy.optimize.minimize(fun=sse, args=additional, x0=np.array([0.1, 0.0001, 1]), bounds = ((0, None), (0, None), (0, None)))

**Comment by andyfaff** [comment]
You're missing a `*` in front of args.

     def sse(params, *args):
         #data, psr = args[0], args[1]
        data, psr = args['dn'], args['r']
        betavar, rho, zeta = params[0], params[1], params[2]
        return sum((data - (rho + zeta*np.exp(betavar*psr))**-1)**2)

**Comment by hermanc1** [comment]
Aaah. Thank you!

**Comment by kirui93** [comment]
How do you get the results of the extra arguments from the solution?

**Comment by JannisErhard** [comment]
This is a question I would also like to know the answer to.

**Comment by mdhaber** [comment]
Could you clarify the question? I think that "extra arguments" refers to whatever is passed in as `args`, and "the solution" is the object returned by `minimize` (e.g. `res`), but I don't know what "results of the extra arguments" means.

In case it helps, I'll clarify that each element of the `args` tuple gets passed into the user's function separately, so:
```python3
 def sse(params, data, psr):
    betavar, rho, zeta = params[0], params[1], params[2]
    return sum((data - (rho + zeta*np.exp(betavar*psr))**-1)**2)
```
would also work.

**Comment by JannisErhard** [comment]
Assume args change during optimisation but they are not the argument of the function that  is optimized. Lets say for example, your function, while producing the value that you want to minimize, also produces some useful information, like an array of strings, from which you could learn something about the effect of your parametrization. Of course you can just call a modified version of the function after optimization, an let it return this extra array. It would be great though, if there was a way to carry this information out, through something, the *args tuple seems to be a natural candidate.

**Comment by mdhaber** [comment]
You can pass in any mutable object into `args` (e.g. an empty array) and store the information there, then look at what's inside when 'minimize` is done. Or, `fun` could be a method of an object, and you could store the information in an attribute of the object.
There is no particular feature of `minimize` for doing this, though. You may also be interested in the `callback` interface, which calls another function, besides `fun`, each iteration.

**Comment by JannisErhard** [comment]
I should clarify, I use the fmin_bfgs method from scipy.optimize. It sure is no problem to hand down an array of strings via `*args` to the function. It is also no problem to manipulate the array on the fly, but in every subsequent call of the function-to-optimize, the strings are "reset", hence no way to finally get them back.

`
    additionalargs = (n_l,n_r,y_lf,y_rf,scaled_data,x_r,x_l,sets_of_reactions,weights,commentary,)
    result = optimize.fmin_bfgs(get_deviation, y, args=additionalargs, epsilon=1.0e-7)
    print(commentary)
` 

this is the function:

`def get_deviation(y,*args):
    devsum = 0.0
    (n_l,n_r,y_lf,y_rf,scaled_data,x_r,x_l,sets_of_reactions,weights,commentary)=args
    print(commentary)
    y_l,y_r= decompose(y,n_l-4,n_r-4)
    y_l = reattach(y_lf[:2],y_l,y_lf[-2:])
    y_r = reattach(y_rf[:2],y_r,y_rf[-2:])
    total_energies = {}
    named_deviations={}
    deviations = {}
    for key in scaled_data:
        total_energies[key] = calculate_total_energies_of_set_pchip(scaled_data[key],x_r,y_r,x_l,y_l)
        named_deviations[key] = calculate_deviations_of_set(total_energies[key],sets_of_reactions[key])
        deviations[key] =  np.empty(len(named_deviations[key]))
        deviations[key] = np.fromiter(named_deviations[key].values(),dtype=float)
        devsum = devsum+np.mean(np.abs(deviations[key]))*weights[key]
    commentary = (' ; '.join(str(key)+' : '+str(np.mean(np.abs(deviations[key][:]))) for key in deviations))
    print(commentary)
    return devsum
`

if commentary = 'word_1' is entered in optimize.fmin_bfgs the first `print (commentary)` will always contain 'word_1', while the second contains the information I require.

**Comment by JannisErhard** [comment]
So I resolved this issue for me, and in case somebody stumbles over this and has as little of a clue es me, I share my solution here. 

The problem is that the scope of the "commentary" string variable ends with the callback functions "get_deviation()"  body. There seems to be no obvious way to generate and hand down a pointer through a shallow copy. 

Instead,  I defined a class, that contains only one variable, a string, in which I can save "commentary" inside the callback function "get_deviation". 

This, I assume, works because the instance of the class has it's own scope.
